pkg/build/builder/dockerutil.go

```diff
@@ -62,15 +73,59 @@ func pushImage(client DockerClient, name string, authConfig docker.AuthConfigura
 	if glog.V(5) {
 		opts.OutputStream = os.Stderr
 	}
-	var err error
+
 	var retriableError = false
 
+	eventWatch, err := eventClient.Watch(
+		kapi.ListOptions{
+			FieldSelector: fields.ParseSelectorOrDie(
+				strings.Join([]string{
+					"involvedObject.kind==" + "ImageStream",
+					"involvedObject.namespace==" + ref.Namespace,
+					"involvedObject.name==" + ref.Name},
+					",")),
+		},
+	)
+	if err != nil {
+		glog.Errorf("Failed to watch for events: %v", err)
+	} else {
+		glog.V(5).Infof("Listening for events on image stream %s", ref.RepositoryName())
+		defer eventWatch.Stop()
+	}
+
 	for retries := 0; retries <= DefaultPushRetryCount; retries++ {
 		err = client.PushImage(opts, authConfig)
 		if err == nil {
 			return nil
 		}
 
+		waitEnd := time.Now().Add(time.Second * 2)
+
+		if eventWatch != nil && strings.Contains(err.Error(), "denied") {
```

Comment:
> This is not a kAPI error.

Yeah, I realize it's a docker error and not a kapi error, just trying to think if this is a strong enough check or if there is a chance it would catch something unexpected.